### ISSUE TITLE
actionunban doesn't work

### ISSUE THREAD

**Issue by JoshuaPettus** [wpkc/fail2ban-action-cloudflare-restv4-bearer-token#2]
For whatever reason your python method doesn't work. the new fail2ban cloudflare.conf file does have a working method but doesn't use the bearer-token. I modified their method with some of yours and that seems to work.  I like the way you handle the "cfkey" in jail.local but theirs also has some handling of ipv6

> actionban = curl -s -o /dev/null -X POST <_cf_api_prms> \
>             -d '{"mode":"block","configuration":{"target":"<cftarget>","value":"<ip>"},"notes":"Fail2Ban <name>"}' \
>             <_cf_api_url>

> actionunban = id=$(curl -s -X GET <_cf_api_prms> \
>                    "<_cf_api_url>?mode=block&configuration_target=<cftarget>&configuration_value=<ip>&page=1&per_page=1&notes=Fail2Ban%%20<name>" \
>                    | { jq -r '.result[0].id' 2>/dev/null || tr -d '\n' | sed -nE 's/^.*"result"\s*:\s*\[\s*\{\s*"id"\s*:\s*"([^"]+)".*$/\1/p'; })
>               if [ -z "$id" ]; then echo "<name>: id for <ip> cannot be found"; exit 0; fi;
>               curl -s -o /dev/null -X DELETE <_cf_api_prms> "<_cf_api_url>/$id"
> 
> _cf_api_url = https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules
> _cf_api_prms = -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json"

> [Init]
> 
> cftarget = ip
> 
> [Init?family=inet6]
> cftarget = ip6

**Comment by wpkc** [comment]
Hi Joshua,
Thanks for letting me know.  This was a quick example I coded when someone in the Fail2Ban repo said they couldn't use Authorization Bearer Tokens with my original filter.  I actually have not used it myself, but now I'm interested in checking it out.

Your code is making use of jq.  I avoided this because jq is not installed by default on many distros, so using it would add an additional dependency.  I did not want to have to convince the Fail2Ban people to add an new dependency.  

Also, I think my use of python -c ... | xargs curl  is far less complex than their solution.  If CloudFlare ever changed the order of parameters of their json responses, their regex would break.

Looking closer at actionunban, I also see some problems...

In the query string, _configuration_target_ should be _configuration.target_, and _configuration_value_ should be _configuration.value_.
[https://api.cloudflare.com/#user-level-firewall-access-rule-list-access-rules](https://api.cloudflare.com/#user-level-firewall-access-rule-list-access-rules)

Also, neither value.  These should be: _configuration.target=ip_ and _configuration_value=<cftarget>_

The only filter set in the query string is _notes=Fail2Ban%%20_  Because of this, your query will return the host listed at the top of the list of all hosts banned by Fail2Ban.  This is fine if there is only one IP address banned at a time, but that's often not the case.  You might end up unbanning a different IP address than you intended to.

Also __cf_api_prms = -H "Authorization: Bearer " -H "Content-Type: application/json"_ is missing your token.  I think it needs to be __cf_api_prms = -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json"_ in order to work at all.  This is assuming that your token is stored in _<cfkey>_

I will try implementing my solution on one of my web sites and see how it goes.

One other thing you might try checking on is the permissions of the token you created for use with this action.  It must have rights to create, delete, and list items in the user level firewall.

Thanks again for the feedback.

**Comment by JoshuaPettus** [comment]
Thank you!

Oh yes, when your initial code didn't work for the unban part, the first thing I looked at was the token permissions.  But then realized that if it can ban, but not unban, than permissions must not be the issue.  That was confirmed when I tried the fail2ban code, but theirs using only a full global api key seemed not idea, hence the mix and match.  (And for someone with my skill level... pray)

**Comment by wpkc** [comment]
Hi Joshua,

I ran both the ban and unban actions from the command line and they worked without any problems.  This (with cfkey replaced with a real value)...
`curl -s -X GET -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" "https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?page=1&per_page=5&mode=block&configuration.target=ip&configuration.value=123.123.123.123&notes=Banned by Fail2Ban&match=all&order=configuration.value&direction=desc" | python -c "import sys, json; print json.load(sys.stdin)['result'][0]['id'];" | xargs -I@@ curl -s -X DELETE -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/@@`

Gave me a response of...
`{
  "result": {
    "id": "9d2.......6e7"
  },
  "success": true,
  "errors": [],
  "messages": []
}`

Now that I confirmed that it is working from the command line, I will configure one of  my web servers to use this action and see if any other problems show up.

**Comment by wpkc** [comment]
Joshua,

I got my first ban/unban action with Fail2Ban using my bearer token.  It worked flawlessly.  I'm not sure what else I can tell you other than reading[ this article](https://www.kazimer.com/fail2ban-cloudflare-action-using-authorization-bearer-token/) to see if it helps.

**Comment by JoshuaPettus** [comment]
Thank you! I'll take a look, try your command line method first when I get a chance.  Maybe that will point to where I got something wrong.

**Comment by wpkc** [comment]
Joshua,

One other thing I want to point out.  It was a bit confusing to me that you found people in the Fail2Ban forum suggesting the following...
```
[Init]

cftarget = ip
 
[Init?family=inet6]
cftarget = ip6
```

I was fairly certain that this is unnecessary to get IPv6 addresses working with any action.  But now I am absolutely certain that this is totally unnecessary, because this morning I received an email from my Fail2Ban setup using my authorization bearer token action showing this...
```
The IP 2600:6c5d:0:3ad0:bd4e:1545:f61e:1c13 has just been banned by Fail2Ban after
4 attempts against apache-auth.
```

**Comment by JoshuaPettus** [comment]
Thank you!  That's good to know!

**Comment by jfl1616** [comment]
As well, I encountered the same issue with the actionunban; however, I managed to resolve it. My Linux system only detected Python3 and not Python. For this one, I created the following alias:
`alias python="Python3"`

Furthermore, it throws the `SyntaxError: invalid syntax` since it requires that I add a parameter for print keyboard.

```
actionunban = curl -s -X GET -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" \
	"https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?page=1&per_page=5&mode=block&configuration.target=ip&configuration.value=<ip>&notes=Banned by Fail2Ban&match=all&order=configuration.value&direction=desc" | \
	python -c "import sys, json; print(json.load(sys.stdin)['result'][0]['id']);" | \
	xargs -I@@ curl -s -X DELETE -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/@@
```

**Comment by sunildmonte** [comment]
Hello, my ban works, but I get the following error in `fail2ban.log` for the unban call, and the offending IP remains blocked in CF:
```
[myjail] action 'cloudflare-restv4-bearer-token': unban 122.172.80.21
 stderr: 'Traceback (most recent call last):'
 stderr: '  File "<string>", line 1, in <module>'
 stderr: '  File "/usr/lib64/python2.7/json/__init__.py", line 291, in load'
 stderr: '    **kw)'
 stderr: '  File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads'
 stderr: '    return _default_decoder.decode(s)'
 stderr: '  File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode'
 stderr: '    obj, end = self.raw_decode(s, idx=_w(s, 0).end())'
 stderr: '  File "/usr/lib64/python2.7/json/decoder.py", line 382, in raw_decode'
 stderr: '    raise ValueError("No JSON object could be decoded")'
 stderr: 'ValueError: No JSON object could be decoded'
 returned successfully 0
```
Any ideas what I'm missing?

**Comment by wpkc** [comment]
I think Fail2Ban uses python3.  This log file shows python2.7 running.  Install python3 and the latest fail2ban.  If you still get the same message, try what the previous commenter did...
https://github.com/wpkc/fail2ban-action-cloudflare-restv4-bearer-token/issues/2#issuecomment-1114061851

**Comment by sunildmonte** [comment]
Hi, I switched to python3 (3.7.10), and got the `SyntaxError: invalid syntax` seen by the previous commenter.
So I then replaced the `actionunban` with the amended version given in that comment, but I still got a python error.
Finally I found the problem: it was the spaces in the `notes=Banned by Fail2Ban` URL parameter. Once I URL-encoded that (with double-% to escape), it works. 
Here is the actionunban:
```
actionunban = curl -s -X GET -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" \
	"https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules?page=1&per_page=5&mode=block&configuration.target=ip&configuration.value=<ip>&notes=Banned%%20by%%20Fail2Ban&match=all&order=configuration.value&direction=desc" | \
	python -c "import sys, json; print(json.load(sys.stdin)['result'][0]['id']);" | \
	xargs -I@@ curl -s -X DELETE -H "Authorization: Bearer <cfkey>" -H "Content-Type: application/json" https://api.cloudflare.com/client/v4/user/firewall/access_rules/rules/@@
```

**Comment by wpkc** [comment]
Glad you figured it out and shared your solution.

There must be some difference between distros, because I have not seen any problems with spaces in the URL.  Adding the encoded spaces is definitely safer though.

I looked at the current Fail2Ban code and found they are using a ridiculously complex `actionunban` and the author doesn't seem to be using the `<cftarget>` variable, but still sets it...
https://github.com/fail2ban/fail2ban/blob/master/config/action.d/cloudflare-token.conf

It also looks like someone took my original API key version of the cloudflare action, made it a bit more complicated, then deleted my credit and put their name on it...
https://github.com/fail2ban/fail2ban/blob/master/config/action.d/cloudflare.conf

Bottom line is that the fail2ban project has it's own implementations of the actions I wrote.  So I should probably just delete my repositories to avoid confusion and parallel development.
